js_modules/dagster-ui/packages/ui-core/src/hooks/useQueryAndLocalStoragePersistedState.tsx

```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+
+import {QueryPersistedDataType, useQueryPersistedState} from './useQueryPersistedState';
+import {useSetStateUpdateCallback} from './useSetStateUpdateCallback';
+
+/**
+ *
+ * Use URL query string as main source of truth with localStorage as the backup if no state is found in the query string
+ * Syncs changes back to localStorage but relies solely on queryString after the initial render
+ * @returns
+ */
+export const useQueryAndLocalStoragePersistedState = <T extends QueryPersistedDataType>(
+  props: Parameters<typeof useQueryPersistedState<T>>[0] & {
+    localStorageKey: string;
+    isEmptyState: (state: T) => boolean;
+  },
+): [T, (setterOrState: React.SetStateAction<T>) => void] => {
+  // Grab state from localStorage as "initialState"
+  const initialState = React.useMemo(() => {
+    try {
+      const value = localStorage.getItem(props.localStorageKey);
+      if (value) {
+        return props.decode?.(JSON.parse(value));
+      }
+    } catch {}
+    return undefined;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [props.localStorageKey]);
+
+  const [state, setter] = useQueryPersistedState(props);
+
+  React.useEffect(() => {
+    // If there was state in local storage and the state is currently empty then use the state
+    // we retrieved from localStorage
+    if (initialState && props.isEmptyState(state)) {
+      setter(initialState);
```

Comment:
I think this useEffect is good, but the `state` returned from this hook below on 42 should probably be `isEmptyState(state) ? initialState : state` so that there isn't one bad render before the previous local storage value is applied. Right now, `state` is incorrect / not the initial state until this useEffect runs.

Reply:
Yeah I can update it to that. FWIW the right solution which would avoid also re-evaluating on every render is kinda clunky so I omitted it.